pkg/apis/core/v1beta1/validation/validation_test.go

```diff
@@ -0,0 +1,259 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"sigs.k8s.io/kubefed/pkg/apis/core/typeconfig"
+	"sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/kubefedctl/enable"
+	"sigs.k8s.io/kubefed/pkg/kubefedctl/options"
+)
+
+func TestValidateFederatedTypeConfig(t *testing.T) {
+	errs := ValidateFederatedTypeConfig(validFederatedTypeConfig(), true)
+	if len(errs) != 0 {
+		t.Errorf("expected success: %v", errs)
+	}
+}
+
+func TestValidateFederatedTypeConfigSpec(t *testing.T) {
+	apiResource := &metav1.APIResource{
+		Group:      "",
+		Version:    "v1",
+		Kind:       "Service",
+		Name:       "services",
+		Namespaced: true,
+	}
+	apiResourceWithGroup := &metav1.APIResource{
+		Group:      "apps",
+		Version:    "v1",
+		Kind:       "Deployment",
+		Name:       "deployments",
+		Namespaced: true,
+	}
+
+	successCases := []*v1beta1.FederatedTypeConfig{
+		federatedTypeConfig(apiResource),
+		federatedTypeConfig(apiResourceWithGroup),
+	}
+	for _, successCase := range successCases {
+		if errs := ValidateFederatedTypeConfigSpec(&successCase.Spec, field.NewPath("spec")); len(errs) != 0 {
+			t.Errorf("expected success: %v", errs)
+		}
+	}
+
+	errorCases := map[string]*v1beta1.FederatedTypeConfig{}
+
+	// Validate required fields
+	fedGroupRequired := validFederatedTypeConfig()
+	fedGroupRequired.Spec.FederatedType.Group = ""
+	errorCases["federatedType.group: Required value"] = fedGroupRequired
+
+	versionRequired := validFederatedTypeConfig()
+	versionRequired.Spec.TargetType.Version = ""
+	errorCases["targetType.version: Required value"] = versionRequired
+
+	kindRequired := validFederatedTypeConfig()
+	kindRequired.Spec.TargetType.Kind = ""
+	errorCases["targetType.kind: Required value"] = kindRequired
+
+	pluralName := validFederatedTypeConfig()
+	pluralName.Spec.TargetType.PluralName = ""
+	errorCases["targetType.pluralName: Required value"] = pluralName
+
+	scope := validFederatedTypeConfig()
+	scope.Spec.TargetType.Scope = ""
+	errorCases["targetType.scope: Required value"] = scope
+
+	propagation := validFederatedTypeConfig()
+	propagation.Spec.Propagation = ""
+	errorCases["spec.propagation: Required value"] = propagation
+
+	// Validate field values
+	validFedGroup := validFederatedTypeConfig()
+	validFedGroup.Spec.FederatedType.Group = "nodomain"
+	errorCases[domainWithAtLeastOneDot] = validFedGroup
+
+	validTargetGroup := validFederatedTypeConfig()
+	validTargetGroup.Spec.TargetType.Group = "invalid#group"
+	errorCases["must start and end with an alphanumeric character"] = validTargetGroup
```

Comment:
it seems the group name already `start and end with an charactor`. I guess you mean `#` should be invalid.